# Load Digits Classification Project

In this project i am going to use a toy dataset from scikit learn dataset in order to solve the classification problem

In [2]:
# Importing the necessary parameters
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
# Importing the dataset
from sklearn.datasets import load_digits
load_digits

<function sklearn.datasets._base.load_digits(*, n_class=10, return_X_y=False, as_frame=False)>

In [4]:
digits_data = load_digits()
load_digits()

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [5]:
digits_df = pd.DataFrame(digits_data["data"], columns=digits_data["feature_names"])
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

In [7]:
digits_df_copy = digits_df.copy()
digits_df_copy.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        5.0        0.0   
3        0.0        0.0        0.0        8.0  ...        9.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0        6.0       13.0       10.0        0.0   
1        0.0        0.0        0.0       11.0       16.0       10.0   
2        0.0        0.0        0.0        3.0       11.0       16.0   
3        0.0        0.0        7.0       13.0       13.0        9.0   
4        0.0        0.0        0.0        2.0       16.0        4.0   

   pixel_7_6  pixel_7_7  
0        0.0        0.0  
1        0.0        0.0  
2        9.0        0.0  
3        0.0        0.0  
4        0.0        0.0  

[5 rows x 64 columns]

In [8]:
digits_df["target"] = pd.DataFrame(digits_data["target"])
digits_df["target"].head()

0    0
1    1
2    2
3    3
4    4
Name: target, dtype: int32

In [9]:
digits_df.isna().sum()

pixel_0_0    0
pixel_0_1    0
pixel_0_2    0
pixel_0_3    0
pixel_0_4    0
            ..
pixel_7_4    0
pixel_7_5    0
pixel_7_6    0
pixel_7_7    0
target       0
Length: 65, dtype: int64

In [10]:
digits_df.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2    pixel_7_3  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      0.206455     0.000556     0.279354     5.557596    12.089037   
std       0.984401     0.023590     0.934302     5.103019     4.374694   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     1.000000    11.000000   
50%       0.000000     0.000000     0.000000     4.000000    13.000000   
75%       0.000000     0.000000     0.000000    10.000000    16.000000   
max      13.000000     1.000000     9.000000    16.000000    16.000000   

         pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7       target  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     11.809126     6.764051     2.067891     0.364496     4.490818  
std       4.933947     5.900623     4.090548     1.860122     2.865304  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      10.000000     0.000000     0.000000     0.000000     2.000000  
50%      14.000000     6.000000     0.000000     0.000000     4.000000  
75%      16.000000    12.000000     2.000000     0.000000     7.000000  
max      16.000000    16.000000    16.000000    16.000000     9.000000  

[8 rows x 65 columns]

## Creating our machine learning model
Done by importing the necessary tools from scikit-learn

In [29]:
np.random.seed(42)
# Importing the necessary tools for model training
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor

# Creating the X and y variable
X = digits_df_copy
y = digits_df["target"]

# Splitting into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

models = {"rf":RandomForestClassifier(random_state=0),
         "reg":RandomForestRegressor(random_state=0)}

results = {}

for model_name, model in models.items():
    model.fit(X_train, y_train)
    results[model_name] = model.score(X_test, y_test)

results

{'rf': 0.9777777777777777, 'reg': 0.8311467879186224}

## Scoring the best performing model

In [13]:
# Importing the classification evaluation metrics
from sklearn.metrics import precision_score, recall_score, f1_score

rf = RandomForestClassifier()
rf.fit(X_train, y_train)
y_preds = rf.predict(X_test)
y_preds[:5]

array([6, 9, 3, 7, 2])

In [14]:
# Calculating the evaluation metrics
np.random.seed(42)
pr_score = precision_score(y_test, y_preds, average=None) 
#Since this is a multiclass classification, average has to be set from the default value of binary to none
np.mean(pr_score)

0.9743063636369385

In [15]:
# Calculating the recall score
r_score = recall_score(y_test, y_preds, average=None)
np.mean(r_score)

0.9731415486891082

In [16]:
# Calculating the f1_score
f1 = f1_score(y_test, y_preds, average=None)
np.mean(f1)

0.9736052008080776

## Hypertuning the best performing model

In [18]:
# Method 1. changing the n_estimators
np.random.seed(42)
for i in range (10,150,10):
    print(f"Trying model with {i}estimators.....")
    clf = RandomForestClassifier(n_estimators=i, random_state=0).fit(X_train, y_train)
    print(f"Model accuracy on test set:{clf.score(X_test, y_test)*100:2f}%")
    

Trying model with 10estimators.....
Model accuracy on test set:95.833333%
Trying model with 20estimators.....
Model accuracy on test set:96.944444%
Trying model with 30estimators.....
Model accuracy on test set:97.777778%
Trying model with 40estimators.....
Model accuracy on test set:97.777778%
Trying model with 50estimators.....
Model accuracy on test set:97.500000%
Trying model with 60estimators.....
Model accuracy on test set:97.222222%
Trying model with 70estimators.....
Model accuracy on test set:97.777778%
Trying model with 80estimators.....
Model accuracy on test set:97.777778%
Trying model with 90estimators.....
Model accuracy on test set:98.055556%
Trying model with 100estimators.....
Model accuracy on test set:97.777778%
Trying model with 110estimators.....
Model accuracy on test set:97.777778%
Trying model with 120estimators.....
Model accuracy on test set:97.777778%
Trying model with 130estimators.....
Model accuracy on test set:98.055556%
Trying model with 140estimators...

From the above test we can see that the n_estimators that produced the best results were 60 to 110
Therefore we are going to use another method of tuning our model

In [27]:
# Method 2
np.random.seed(42)
from sklearn.model_selection import GridSearchCV

grid_1 = {"n_estimators":np.arange(60,150,10),
         "max_depth":np.arange(0,10,1),
         "max_features":np.arange(0.1,1,0.1)}

# Creating another copy of the original dataframe
df_copy_2 =digits_df.copy()

# Shuffling the new copy of the dataframe
df_shuffled = df_copy_2.sample(frac=1)
df_shuffled.head()

# Creating the X and y variable
X = df_shuffled.drop("target", axis=1)
y = df_shuffled["target"]

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size =0.2, random_state=0)

# Defining the classifier model
clf_2 = RandomForestClassifier(n_jobs=-1, random_state = 0)

gs_clf_2 = GridSearchCV(estimator=clf_2,
                       param_grid=grid_1,
                       cv = 5,
                       verbose=2)
gs_clf_2.fit(X_train, y_train)
gs_clf_2.best_params_

Fitting 5 folds for each of 810 candidates, totalling 4050 fits
[CV] END .....max_depth=0, max_features=0.1, n_estimators=60; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=60; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=60; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=60; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=60; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=70; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=70; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=70; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=70; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=70; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, n_estimators=80; total time=   0.0s
[CV] END .....max_depth=0, max_features=0.1, 

C:\Users\Admin\Desktop\PROJECTS\sample_project_1\env\Lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
405 fits failed out of a total of 4050.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
405 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Admin\Desktop\PROJECTS\sample_project_1\env\Lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Admin\Desktop\PROJECTS\sample_project_1\env\Lib\site-packages\sklearn\ensemble\_forest.py", line 340, in fit
    self._validate_params()
  File "C:\Users\Admin\Desktop\PROJECTS\sample_project_1\env\Lib\site-packages\sklearn\base.py",

{'max_depth': 9, 'max_features': 0.1, 'n_estimators': 140}

In [30]:
clf_3 = RandomForestClassifier(n_estimators=140, random_state=0)
clf_3.fit(X_train, y_train)
clf_3.score(X_test, y_test)

0.9805555555555555

In [31]:
import pickle
pickle.dump(clf_3, open("clf_3_random_forest_model.pkl","wb"))